msgpack-core/src/main/java/org/msgpack/value/impl/ImmutableArrayValueImpl.java

```diff
@@ -0,0 +1,210 @@
+//
+// MessagePack for Java
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+//
+package org.msgpack.value.impl;
+
+import org.msgpack.core.MessagePacker;
+import org.msgpack.value.Value;
+import org.msgpack.value.ValueType;
+import org.msgpack.value.ArrayValue;
+import org.msgpack.value.ImmutableArrayValue;
+
+import java.util.List;
+import java.util.AbstractList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.io.IOException;
+
+
+/**
+ * {@code ImmutableArrayValueImpl} Implements {@code ImmutableArrayValue} using a {@code Value[]} field.
+ *
+ * @see  org.msgpack.value.IntegerValue
+ */
+public class ImmutableArrayValueImpl extends AbstractImmutableValue implements ImmutableArrayValue {
+    private static ImmutableArrayValueImpl EMPTY = new ImmutableArrayValueImpl(new Value[0]);
+
+    public static ImmutableArrayValue empty() {
+        return EMPTY;
+    }
+
+    private final Value[] array;
+
+    public ImmutableArrayValueImpl(Value[] array) {
+        this.array = array;
+    }
+
+    @Override
+    public ValueType getValueType() {
+        return ValueType.ARRAY;
+    }
+
+    @Override
+    public ImmutableArrayValue immutableValue() {
+        return this;
+    }
+
+    @Override
+    public ImmutableArrayValue asArrayValue() {
+        return this;
+    }
+
+    @Override
+    public int size() {
+        return array.length;
+    }
+
+    @Override
+    public Value get(int index) {
+        return array[index];
+    }
+
+    @Override
+    public Value getOrNilValue(int index) {
+        if (array.length < index && index >= 0) {
```

Comment:
This condition seems to be backwards. I think it should be `index < array.length`